File: crop_avatars.py
import os
import sys
import numpy as np
from PIL import Image, ImageDraw

def crop_avatars(image_path, output_dir="app/src/main/res/drawable"):
    if not os.path.exists(image_path):
        print(f"Error: Input image file '{image_path}' not found.")
        print("Please upload the file to the workspace first!")
        return

    # Load image and convert to RGBA
    try:
        img = Image.open(image_path).convert("RGBA")
    except Exception as e:
        print(f"Error opening image: {e}")
        return

    os.makedirs(output_dir, exist_ok=True)
    W, H = img.size
    print(f"Original Image Size: {W}x{H}")

    # We have 2 rows and 4 columns of avatars
    rows = 2
    cols = 4
    cell_w = W // cols
    cell_h = H // rows

    avatar_count = 0

    for r in range(rows):
        for c in range(cols):
            avatar_count += 1
            # Step 1: Crop the rough cell
            left = c * cell_w
            top = r * cell_h
            right = (c + 1) * cell_w
            bottom = (r + 1) * cell_h
            cell = img.crop((left, top, right, bottom))

            # Step 2: Find bounding box of non-white pixels
            # Convert to numpy array to find non-white pixels quickly
            data = np.array(cell)
            # Threshold to identify non-white/non-transparent background
            # Background is white (255, 255, 255)
            r_channel = data[:, :, 0]
            g_channel = data[:, :, 1]
            b_channel = data[:, :, 2]
            alpha_channel = data[:, :, 3]

            # Pixels that are not white (with a small tolerance) and are not transparent
            non_white_mask = (r_channel < 240) | (g_channel < 240) | (b_channel < 240)
            if alpha_channel.any():
                non_white_mask = non_white_mask & (alpha_channel > 50)

            # Find coordinates of non-white pixels
            y_indices, x_indices = np.where(non_white_mask)

            if len(x_indices) == 0 or len(y_indices) == 0:
                print(f"Warning: Could not find avatar in row {r+1}, col {c+1}")
                continue

            min_x, max_x = x_indices.min(), x_indices.max()
            min_y, max_y = y_indices.min(), y_indices.max()

            # The exact bounding box of the circular avatar (with gold border)
            width = max_x - min_x
            height = max_y - min_y
            size = max(width, height)

            # Center of the detected circle
            cx = min_x + width / 2
            cy = min_y + height / 2
            radius = size / 2

            # Let's add a tiny margin of 2 pixels to make sure the outer edge of the gold border isn't clipped
            radius += 2
            size = int(radius * 2)

            # Crop the circle area with precise coordinates
            crop_box = (
                int(cx - radius),
                int(cy - radius),
                int(cx + radius),
                int(cy + radius)
            )
            cropped_avatar = cell.crop(crop_box)

            # Create circular transparent mask
            mask = Image.new("L", (size, size), 0)
            draw = ImageDraw.Draw(mask)
            draw.ellipse((0, 0, size, size), fill=255)

            # Apply circular mask
            output_avatar = Image.new("RGBA", (size, size), (0, 0, 0, 0))
            output_avatar.paste(cropped_avatar, (0, 0), mask=mask)

            # Save the cropped avatar
            output_filename = f"ic_avatar_{avatar_count}.png"
            output_filepath = os.path.join(output_dir, output_filename)
            output_avatar.save(output_filepath, "PNG")
            print(f"Successfully saved Avatar {avatar_count} -> {output_filepath} (Size: {size}x{size})")

    print("\nAll 8 avatars cropped and saved successfully in the drawables directory!")

if __name__ == "__main__":
    img_file = "avatars.png"
    if len(sys.argv) > 1:
        img_file = sys.argv[1]
    crop_avatars(img_file)

File: CropAvatars.java
import java.io.File;
import java.io.IOException;
import java.awt.*;
import java.awt.image.BufferedImage;
import javax.imageio.ImageIO;

public class CropAvatars {
    public static void main(String[] args) {
        String inputPath = "avatars.png";
        if (args.length > 0) {
            inputPath = args[0];
        }
        File inputFile = new File(inputPath);
        if (!inputFile.exists()) {
            System.out.println("Error: Input file '" + inputPath + "' does not exist.");
            System.out.println("Please upload your image and name it 'avatars.png' in the root directory.");
            return;
        }

        try {
            BufferedImage img = ImageIO.read(inputFile);
            int W = img.getWidth();
            int H = img.getHeight();
            System.out.println("Loaded image successfully. Dimensions: " + W + "x" + H);

            int rows = 2;
            int cols = 4;
            int cellW = W / cols;
            int cellH = H / rows;

            File outputDir = new File("app/src/main/res/drawable");
            if (!outputDir.exists()) {
                outputDir.mkdirs();
            }

            int count = 1;
            for (int r = 0; r < rows; r++) {
                for (int c = 0; c < cols; c++) {
                    int left = c * cellW;
                    int top = r * cellH;

                    // Scan cell to find the exact bounding box of the circular gold border
                    int minX = cellW, maxX = 0;
                    int minY = cellH, maxY = 0;
                    boolean found = false;

                    for (int x = 0; x < cellW; x++) {
                        for (int y = 0; y < cellH; y++) {
                            int px = img.getRGB(left + x, top + y);
                            int alpha = (px >> 24) & 0xff;
                            int red = (px >> 16) & 0xff;
                            int green = (px >> 8) & 0xff;
                            int blue = px & 0xff;

                            // Ignore transparent background or white background (with tolerance of 240 out of 255)
                            if (alpha > 50 && (red < 240 || green < 240 || blue < 240)) {
                                if (x < minX) minX = x;
                                if (x > maxX) maxX = x;
                                if (y < minY) minY = y;
                                if (y > maxY) maxY = y;
                                found = true;
                            }
                        }
                    }

                    if (!found) {
                        System.out.println("Warning: No avatar found in Row " + (r + 1) + ", Col " + (c + 1));
                        continue;
                    }

                    // Calculate width, height, and center coordinates
                    int width = maxX - minX;
                    int height = maxY - minY;
                    int size = Math.max(width, height);

                    double cx = minX + width / 2.0;
                    double cy = minY + height / 2.0;
                    
                    // Add 2 pixels of margin to make sure we don't clip the outer edge of the gold circle
                    double radius = size / 2.0 + 2.0; 
                    int outSize = (int)(radius * 2);

                    // Create the transparent cropped image
                    BufferedImage dest = new BufferedImage(outSize, outSize, BufferedImage.TYPE_INT_ARGB);
                    Graphics2D g2 = dest.createGraphics();
                    
                    // Apply high-quality antialiasing settings for perfect round edges
                    g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                    g2.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
                    
                    // Set alpha transparency
                    g2.setComposite(AlphaComposite.Clear);
                    g2.fillRect(0, 0, outSize, outSize);
                    g2.setComposite(AlphaComposite.Src);

                    // Create circular clip path
                    g2.setClip(new java.awt.geom.Ellipse2D.Double(0, 0, outSize, outSize));

                    // Coordinates to grab from the original image
                    int sourceX1 = (int)(left + cx - radius);
                    int sourceY1 = (int)(top + cy - radius);
                    int sourceX2 = (int)(left + cx + radius);
                    int sourceY2 = (int)(top + cy + radius);

                    g2.drawImage(img, 
                                 0, 0, outSize, outSize, 
                                 sourceX1, sourceY1, sourceX2, sourceY2, 
                                 null);
                    
                    g2.dispose();

                    // Save inside resource drawables
                    File outFile = new File(outputDir, "ic_avatar_" + count + ".png");
                    ImageIO.write(dest, "PNG", outFile);
                    System.out.println("Processed Avatar " + count + " -> " + outFile.getPath() + " (" + outSize + "x" + outSize + ")");
                    count++;
                }
            }
            System.out.println("\nSuccess! Perfectly cropped and saved " + (count - 1) + " transparent circular avatars.");
        } catch (IOException e) {
            System.out.println("Error reading/writing image: " + e.getMessage());
        }
    }
}
